### PR TITLE
10.0 pos multi session sync 28

### PR DESCRIPTION
**Pull request by KolushovAlexandr**

### CODE REVIEW

File: pos_longpolling/static/src/js/pos_longpolling.js [hunk 1]
```diff
@@ -45,8 +45,13 @@ odoo.define('pos_longpolling', function(require){
                 if(!self.stop){
                     self.poll();
                 }
-
+                //new condition: change longpolling icon status if needed
```
Comment (yelizariev) [inline pos_longpolling/static/src/js/pos_longpolling.js]: Why you say "new"? Comments have to make sense not only in PR

File: pos_longpolling/static/src/js/pos_longpolling.js [hunk 2]
```diff
@@ -45,8 +45,13 @@ odoo.define('pos_longpolling', function(require){
...
+                if (!self.longpolling_connection.status) {
```
Comment (yelizariev) [inline pos_longpolling/static/src/js/pos_longpolling.js]: move this check to network_is_on and just call network_is_on here

File: pos_longpolling/static/src/js/pos_longpolling.js [hunk 3]
```diff
@@ -45,8 +45,13 @@ odoo.define('pos_longpolling', function(require){
...
+                    self.longpolling_connection.network_is_on();
+                }
             }, function(unused, e) {
+                //new line: change longpolling icon status
```
Comment (yelizariev) [inline pos_longpolling/static/src/js/pos_longpolling.js]: When developer read the code first time he doesn't understand what is "new". All lines were new at some moment

File: pos_longpolling/static/src/js/pos_longpolling.js [hunk 4]
```diff
@@ -164,6 +169,7 @@ odoo.define('pos_longpolling', function(require){
             this.status = false;
             // Is the message "PONG" received from the server
             this.response_status = false;
+            this.connection_attempt = false;
```
Comment (yelizariev) [inline pos_longpolling/static/src/js/pos_longpolling.js]: What is connection_attempt?

File: pos_multi_session/static/src/js/pos_multi_session.js [hunk 5]
```diff
@@ -90,9 +90,14 @@ odoo.define('pos_multi_session', function(require){
...
                     self.sync_bus = self.get_bus('sync_server');
                     self.get_bus('sync_server').start();
+                    if (!self.config.autostart_longpolling) {
+                        self.bus.activated = false;
+                        self.bus.start();
+                    }
                 } else {
                     self.sync_bus = self.get_bus();
                     if (!self.config.autostart_longpolling) {
+                        self.sync_bus.activated = false;
```
Comment (yelizariev) [inline pos_multi_session/static/src/js/pos_multi_session.js]: autostart_longpolling is only for main server!

File: pos_multi_session/static/src/js/pos_multi_session.js [hunk 6]
```diff
@@ -90,9 +90,14 @@ odoo.define('pos_multi_session', function(require){
                     self.get_bus('sync_server').add_channel_callback(channel_name, callback, self);
                     self.sync_bus = self.get_bus('sync_server');
                     self.get_bus('sync_server').start();
+                    if (!self.config.autostart_longpolling) {
+                        self.bus.activated = false;
+                        self.bus.start();
```
Comment (yelizariev) [inline pos_multi_session/static/src/js/pos_multi_session.js]: Why do you start main longpolling if it's switched off in settings and another sync server is used?

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_longpolling/static/src/js/pos_longpolling.js (hunk 1 above)
Comment (KolushovAlexandr) [reply pos_longpolling/static/src/js/pos_longpolling.js]: @yelizariev all this 'new' comments is beneath this one: // function is copy-pasted from bus.js but the line below defines a custom server address

File: pos_multi_session/static/src/js/pos_multi_session.js (hunk 6 above)
Comment (KolushovAlexandr) [reply pos_multi_session/static/src/js/pos_multi_session.js]: @yelizariev it's not possible to turn off the main longpolling, We can only turn off its autostart in longpolling.js

File: pos_multi_session/static/src/js/pos_multi_session.js (hunk 5 above)
Comment (KolushovAlexandr) [reply pos_multi_session/static/src/js/pos_multi_session.js]: @yelizariev Yes, that is. self.get_bus(); returns pos.bus - main server

File: pos_multi_session/static/src/js/pos_multi_session.js (hunk 6 above)
Comment (yelizariev) [reply pos_multi_session/static/src/js/pos_multi_session.js]: Don't turn it off. Just don't start it

Comment (yelizariev) [reply pos_multi_session/static/src/js/pos_multi_session.js]: Just remove this block

File: pos_multi_session/static/src/js/pos_multi_session.js (hunk 5 above)
Comment (yelizariev) [reply pos_multi_session/static/src/js/pos_multi_session.js]: ok

**Comment by yelizariev** [comment]
- [x] ``....activated = false;`` -- must not be handled in pos_lonpolling (add new argument or new method, if needed)

File: pos_longpolling/static/src/js/pos_longpolling.js [hunk 7]
```diff
@@ -294,6 +291,9 @@ odoo.define('pos_longpolling', function(require){
         start: function(){
             this._super();
             var self = this;
+            if (!this.pos.config.autostart_longpolling) {
+                $('.serv_primary').remove();
+            }
```
Comment (yelizariev) [inline pos_longpolling/static/src/js/pos_longpolling.js]: Don't remove and make it gray instead.

File: pos_longpolling/static/src/js/pos_longpolling.js [hunk 8]
```diff
@@ -73,9 +73,6 @@ odoo.define('pos_longpolling', function(require){
             }
         },
         start: function(){
-            if (this.activated){
-                return;
-            }
```
Comment (yelizariev) [inline pos_longpolling/static/src/js/pos_longpolling.js]: Wrong. It may lead to double initialization

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_longpolling/static/src/js/pos_longpolling.js (hunk 8 above)
Comment (yelizariev) [reply pos_longpolling/static/src/js/pos_longpolling.js]: Or explain why we need it and why it's safe

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_longpolling/static/src/js/pos_longpolling.js [hunk 9]
```diff
@@ -291,11 +294,14 @@ odoo.define('pos_longpolling', function(require){
...
             this._super();
             var self = this;
-            if (!this.pos.config.autostart_longpolling) {
-                $('.serv_primary').remove();
+            var selector = '.serv_primary';
+            var element = $(selector);
+            if (!this.pos.bus.channels.length) {
+                element.removeClass('oe_red');
+                element.removeClass('oe_green');
+                element.addClass('oe_gray');
```
Comment (yelizariev) [inline pos_longpolling/static/src/js/pos_longpolling.js]: You mix logic and UI. Put this code to qweb or renderElement

**Review by yelizariev — CHANGES_REQUESTED** [review]

File: pos_longpolling/static/src/js/pos_longpolling.js [hunk 10]
```diff
@@ -238,12 +241,20 @@ odoo.define('pos_longpolling', function(require){
...
+                } else {
+                    element.removeClass('oe_green');
+                    element.removeClass('oe_gray');
+                    element.addClass('oe_red');
+                }
             } else {
+                element.removeClass('oe_red');
                 element.removeClass('oe_green');
-                element.addClass('oe_red');
+                element.addClass('oe_gray');
```
Comment (yelizariev) [inline pos_longpolling/static/src/js/pos_longpolling.js]: It must be gray when polling doesn't work (longpolling requests are not sent). It's not the same as empty channels list

Comment (yelizariev) [reply pos_longpolling/static/src/js/pos_longpolling.js]: You can simplify code:
```

//compute new_class
...
// set class
element.removeClass('oe_red oe_green oe_gray').addClass(new_class)
```

**Review by yelizariev — CHANGES_REQUESTED** [review]

**Pull request merged**